File: desktop/Protocol/ErrorPayload.cs
using System.Text;

namespace CTRL.Desktop.Protocol;

public sealed record ErrorPayload(byte Code, byte Severity, string Message);

public static class ErrorPayloadCodec
{
    public const byte CodeProtocolVersionMismatch = 0x01;
    public const byte CodeAuthFailed = 0x02;
    public const byte CodeNotAuthenticated = 0x03;
    public const byte CodeDeviceLimit = 0x04;
    public const byte CodePayloadTooLarge = 0x05;
    public const byte CodeInvalidMessage = 0x06;
    public const byte CodeUnsupportedMessage = 0x07;
    public const byte CodeForbidden = 0x08;
    public const byte CodeServerShutdown = 0x09;
    public const byte CodeInternal = 0xFF;

    public const byte SeverityInfo = 0x00;
    public const byte SeverityWarn = 0x01;
    public const byte SeverityFatal = 0x02;

    public const int MaxMessageLength = 1024;

    public static byte[] Encode(ErrorPayload payload)
    {
        if (!IsKnownCode(payload.Code))
            throw new ProtocolException("Unknown ERROR code.");
        if (!IsKnownSeverity(payload.Severity))
            throw new ProtocolException("Unknown ERROR severity.");

        var messageBytes = StrictUtf8.GetBytes(payload.Message);
        if (messageBytes.Length > MaxMessageLength)
            throw new ProtocolException("ERROR message exceeds 1024 bytes.");

        var writer = new PayloadWriter();
        writer.WriteUInt8(payload.Code);
        writer.WriteUInt8(payload.Severity);
        writer.WriteUInt16((ushort)messageBytes.Length);
        writer.WriteBytes(messageBytes);
        return writer.ToArray();
    }

    public static ErrorPayload Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var code = reader.ReadUInt8();
        var severity = reader.ReadUInt8();
        var messageLength = reader.ReadUInt16();
        if (messageLength > MaxMessageLength)
            throw new ProtocolException("ERROR message exceeds 1024 bytes.");
        var messageBytes = reader.ReadBytes(messageLength);
        reader.ExpectEnd();

        if (!IsKnownCode(code))
            throw new ProtocolException("Unknown ERROR code.");
        if (!IsKnownSeverity(severity))
            throw new ProtocolException("Unknown ERROR severity.");

        string message;
        try
        {
            message = StrictUtf8.GetString(messageBytes);
        }
        catch (DecoderFallbackException)
        {
            throw new ProtocolException("Invalid UTF-8 message.");
        }

        return new ErrorPayload(code, severity, message);
    }

    private static bool IsKnownCode(byte code) =>
        code is CodeProtocolVersionMismatch or CodeAuthFailed or
            CodeNotAuthenticated or CodeDeviceLimit or CodePayloadTooLarge or
            CodeInvalidMessage or CodeUnsupportedMessage or CodeForbidden or
            CodeServerShutdown or CodeInternal;

    private static bool IsKnownSeverity(byte severity) =>
        severity is SeverityInfo or SeverityWarn or SeverityFatal;

    private static readonly UTF8Encoding StrictUtf8 =
        new(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
}

File: mobile/lib/protocol/error_payload.dart
import 'dart:convert';
import 'dart:typed_data';

import 'frame.dart';
import 'payload_codec.dart';

const int errorCodeProtocolVersionMismatch = 0x01;
const int errorCodeAuthFailed = 0x02;
const int errorCodeNotAuthenticated = 0x03;
const int errorCodeDeviceLimit = 0x04;
const int errorCodePayloadTooLarge = 0x05;
const int errorCodeInvalidMessage = 0x06;
const int errorCodeUnsupportedMessage = 0x07;
const int errorCodeForbidden = 0x08;
const int errorCodeServerShutdown = 0x09;
const int errorCodeInternal = 0xFF;

const int errorSeverityInfo = 0x00;
const int errorSeverityWarn = 0x01;
const int errorSeverityFatal = 0x02;

const int errorMaxMessageLength = 1024;

class ErrorPayload {
  const ErrorPayload({
    required this.code,
    required this.severity,
    required this.message,
  });

  final int code;
  final int severity;
  final String message;

  @override
  bool operator ==(Object other) =>
      other is ErrorPayload &&
      code == other.code &&
      severity == other.severity &&
      message == other.message;

  @override
  int get hashCode => Object.hash(code, severity, message);
}

class ErrorPayloadCodec {
  static Uint8List encode(ErrorPayload payload) {
    if (!_isKnownCode(payload.code)) {
      throw const ProtocolException('Unknown ERROR code.');
    }
    if (!_isKnownSeverity(payload.severity)) {
      throw const ProtocolException('Unknown ERROR severity.');
    }

    final messageBytes = utf8.encode(payload.message);
    if (messageBytes.length > errorMaxMessageLength) {
      throw const ProtocolException('ERROR message exceeds 1024 bytes.');
    }

    final writer = PayloadWriter()
      ..writeUInt8(payload.code)
      ..writeUInt8(payload.severity)
      ..writeUInt16(messageBytes.length)
      ..writeBytes(messageBytes);
    return writer.toBytes();
  }

  static ErrorPayload decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final code = reader.readUInt8();
    final severity = reader.readUInt8();
    final messageLength = reader.readUInt16();
    if (messageLength > errorMaxMessageLength) {
      throw const ProtocolException('ERROR message exceeds 1024 bytes.');
    }
    final message = reader.readStringOfLength(messageLength);
    reader.expectEnd();

    if (!_isKnownCode(code)) {
      throw const ProtocolException('Unknown ERROR code.');
    }
    if (!_isKnownSeverity(severity)) {
      throw const ProtocolException('Unknown ERROR severity.');
    }

    return ErrorPayload(code: code, severity: severity, message: message);
  }

  static bool _isKnownCode(int code) =>
      code == errorCodeProtocolVersionMismatch ||
      code == errorCodeAuthFailed ||
      code == errorCodeNotAuthenticated ||
      code == errorCodeDeviceLimit ||
      code == errorCodePayloadTooLarge ||
      code == errorCodeInvalidMessage ||
      code == errorCodeUnsupportedMessage ||
      code == errorCodeForbidden ||
      code == errorCodeServerShutdown ||
      code == errorCodeInternal;

  static bool _isKnownSeverity(int severity) => severity <= errorSeverityFatal;
}
